pypeit/edgetrace.py

```diff
@@ -129,6 +130,102 @@
 from pypeit.spectrographs import slitmask
 
 
+class SlitTraceSet(DataContainer):
+    """
+    Defines a generic class for holding and manipulating image traces
+    organized into left-right slit pairs.
+
+    Instantiation arguments map directly to the object
+    :attr:`datamodel`.
+    """
+    # Define the data model
+    datamodel = {'spectrograph': dict(otype=str, descr='Spectrograph used to take the data.'),
```

Comment:
I suggest you add 2 additional items:

1. datamodel_name = dict(otype=str, descr='Name of the DataModel')
2. dm_version = dict(otype=str, descr='Version of the DataModel')

you can choose better names for these attributes.
Maybe the first item is handled by the extension name,
but I'd still like a way to see it trivially.

Reply:
I like the idea of versioning the datamodel, but I'd rather the version not be an element of the datamodel, if that makes sense.  And the datamodel name should be the class.  So here's my proposal:  I'll add a `version` attribute to `DataContainer` and edit the `__setattr__` function so that it's immutable.  Something like this example:

```
class Test:
    version = '1.0.0'
    def __init__(self):
        self.test = 2
    def __setattr__(self, item, value):
        if item == 'version':
            raise TypeError('Internal version does not support assignment.')
        return dict.__setattr__(self, item, value)
```
Then I'll add the class name and the version to the header(s).

Reply:
Okay.  Added this.  Have a look at the changes when I push.  E.g., the trace master now has the following in the header items in the SLITS extension:

```
...
VERSPYT = '3.7.4   '           / Python version
VERSNPY = '1.18.0  '           / Numpy version
VERSSCI = '1.4.1   '           / Scipy version
VERSAST = '4.0     '           / Astropy version
VERSSKL = '0.21.3  '           / Scikit-learn version
VERSPYP = '0.12.2dev'          / PypeIt version
DATE    = '2020-01-14'         / UTC date created
DATAMOD = 'SlitTraceSet'       / Datamodel class
DATAVER = '1.0.0   '           / Datamodel version
...
```